src/super_gradients/module_interfaces/exportable_detector.py

```diff
@@ -0,0 +1,396 @@
+import abc
+import copy
+import dataclasses
+from typing import Any
+from typing import Union, Optional, List, Tuple
+
+import onnxsim
+import torch
+from torch import nn, Tensor
+from torch.utils.data import DataLoader
+
+from super_gradients.common.abstractions.abstract_logger import get_logger
+from super_gradients.conversion.conversion_enums import ExportTargetBackend, ExportQuantizationMode, DetectionOutputFormatMode
+from super_gradients.conversion.onnx.nms import attach_onnx_nms
+from super_gradients.conversion.preprocessing_modules import CastTensorTo
+from super_gradients.conversion.tensorrt.nms import attach_tensorrt_nms
+from super_gradients.training.utils.export_utils import infer_format_from_file_name, infer_image_shape_from_model, infer_image_input_channels
+from super_gradients.training.utils.quantization.fix_pytorch_quantization_modules import patch_pytorch_quantization_modules_if_needed
+from super_gradients.training.utils.utils import infer_model_device
+
+logger = get_logger(__name__)
+
+__all__ = ["ExportableObjectDetectionModel", "AbstractObjectDetectionDecodingModule", "ModelExportResult"]
+
+
+class AbstractObjectDetectionDecodingModule(nn.Module):
+    @abc.abstractmethod
+    def forward(self, predictions: Any) -> Tuple[Tensor, Tensor]:
+        """
+
+        :param predictions:
+        :return:
+        """
+        ...
+
+    def get_output_names(self) -> List[str]:
+        """
+        Returns the names of the outputs of the module.
+
+        :return:
+        """
+        return ["pre_nms_bboxes_xyxy", "pre_nms_scores"]
+
+    @abc.abstractmethod
+    def get_num_pre_nms_predictions(self) -> int:
+        ...
+
+
+@dataclasses.dataclass
+class ModelExportResult:
+    input_image_channels: int
+    input_image_dtype: torch.dtype
+    input_image_shape: Tuple[int, int]
+
+    engine: ExportTargetBackend
+    quantization_mode: Optional[ExportQuantizationMode]
+
+    output: str
+    output_predictions_format: DetectionOutputFormatMode
+
+
+class ExportableObjectDetectionModel:
+    def get_decoding_module(self, num_pre_nms_predictions: int, **kwargs) -> AbstractObjectDetectionDecodingModule:
+        """
+        Gets the decoding module for the object detection model.
+        This method must be implemented by the derived class and should return
+        an instance of AbstractObjectDetectionDecodingModule that would take raw models' outputs and
+        convert them to a tuple of two tensors (boxes, scores):
+         - boxes: [B, N, 4] - All predicted boxes in (x1, y1, x2, y2) format.
+         - scores: [B, N, C] - All predicted scores ([0..1] range) for each box and class.
+        :return: An instance of AbstractObjectDetectionDecodingModule
+        """
+        raise NotImplementedError()
+
+    def get_preprocessing_callback(self, **kwargs) -> Optional[nn.Module]:
+        raise NotImplementedError(f"get_preprocessing_callback is not implemented for class {self.__class__.__name__}.")
```

Comment:
Is there a reason why we go for `raise NotImplementedError()` instead of making it abstract ? To be more precise;

- Do we want to support instantiation of base `ExportableObjectDetectionModel` class ? 
- Is there any situation where we may want to inherit from `ExportableObjectDetectionModel` without implementing `get_decoding_module`  or `get_preprocessing_callback` ?

Reply:
> Do we want to support instantiation of base ExportableObjectDetectionModel class ?
Nope

> Is there any situation where we may want to inherit from ExportableObjectDetectionModel without implementing get_decoding_module or get_preprocessing_callback ?
As of now - no, I cannot think about such cases.

> Is there a reason why we go for raise NotImplementedError() instead of making it abstract ? To be more precise

You mean subclassing from `abc` and marking methods `abstractmethod`? Maybe, need to test whether this would work fine with inheritance chain.